src/err.rs

```diff
@@ -0,0 +1,26 @@
+// Copyright 2016 Jonathan 'theJPster' Pallant.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+//! Implements types influenced by `std::io::Result` for indicating success or failure
+//! of a routine.
+
+use std::result;
+
+/// Used by routines that return a certain type if they pass, or
+/// wish to indicate failure without any sense of why they failed.
+pub type Result<T> = result::Result<T, ()>;
+
+/// Used by routines that wish to indicate success or failure, with
+/// no associated value either way.
+pub type EmptyResult = Result<()>;
```

Comment:
Could you call the enum RingError and give it a single member called
Unspecified? It would give you options in the future for being more
specific with errors and without requiring a second breaking change to the
API.

I agree the result::Result/Result confusion is a bit of a mess. If you get
the API right first time (or at least this time) then aliasing the return
type in this way could be argued as unnecessary.

On 8 Aug 2016 02:04, "Brian Smith" <notifications@github.com> wrote:

In src/err.rs
<https://github.com/briansmith/ring/pull/250#discussion_r73813197>:

> +// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
> +// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
> +// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
> +
> +//! Implements types influenced by `std::io::Result` for indicating success or failure
> +//! of a routine.
> +
> +use std::result;
> +
> +/// Used by routines that return a certain type if they pass, or
> +/// wish to indicate failure without any sense of why they failed.
> +pub type Result<T> = result::Result<T, ()>;
> +
> +/// Used by routines that wish to indicate success or failure, with
> +/// no associated value either way.
> +pub type EmptyResult = Result<()>;

The main goal is to make it possible to implement the From trait to convert
a *ring* error into an application-specific error.

Instead of this, I think we need something like: enum UnspecifiedError; and
then change the return types of functions from Result<R, ()> to Result<R,
UnspecifiedError>. Then the application can implement
From<ring::UnspecifiedError> or From<Result<R, ring::UnspecifiedError>> to
get the conversion as described in https://doc.rust-lang.org/
book/error-handling.html#composing-custom-error-types.

We could also implement std::error::Error for UnspecifiedError to get the
"default" conversion as specified here: https://doc.rust-lang.org/
book/error-handling.html#the-from-trait.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/briansmith/ring/pull/250/files/0b98364802124e49957ff6203bd45677fdc2fa80#r73813197>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AA6lj9gNeuqHcKw0a35K0fXTjXfAwMEKks5qdoCRgaJpZM4JOB7R>
.